Java sprint 8 v1.0

1EVILGUN1 opened the pull request:
В классе Task и его наследниках были добавлены переменные startTime, duration, endTime. В классе FileBackedTaskManager добавлены поля для времени старта и продолжительности. 2 forEach переделаны под Stream API в классе InMemoryTaskManager. Также добавлен TreeSet prioritizedTasks с методо для поиска пересечений во времени addPriorityTask

1EVILGUN1 committed:
Java sprint 8 v1.0

В классе Task и его наследниках были добавлены переменные startTime, duration, endTime. В классе FileBackedTaskManager добавлены поля для времени старта и продолжительности. 2 forEach переделаны под Stream API в классе InMemoryTaskManager

1EVILGUN1 committed:
Java sprint 8 v1.01

Подкорректировал код

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Task.java:
```diff
@@ -1,24 +1,37 @@
-    public Task(String name, Status status, String description) {
+    private Instant startTime; // LocalDateTime
+    private Duration duration; // минуты или Duration
+    private Instant endTime;
```
Обрати внимание что в задании не говорится что нужно добавлять поле  endTime; для Task
Там говорится: 
"getEndTime() — дата и время завершения задачи, которые рассчитываются исходя из startTime и duration."

YazykovMikhail replied:
Стоит добавить метод

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Task.java:
```diff
@@ -1,24 +1,37 @@
     private Status status;
 
-    public Task(String name, Status status, String description) {
+    private Instant startTime; // LocalDateTime
```
Почему Instant а не LocalDateTime, я бы выбрал LocalDateTime

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/model/Epic.java:
```diff
@@ -1,18 +1,20 @@
 public class Epic extends Task {
 
     private final List<Integer> subtaskIds = new ArrayList<>();
 
```
а вот тут как раз должнол быть поле:
	private LocalDateTime endTime;
Потому что время выполнения эпика зависит от его сабтасок и высчитывается аналогично определению статуса Эпика

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/InMemoryTaskManager.java:
```diff
@@ -232,4 +240,16 @@ private void updateEpicStatus(int epicId) {// обновление статус
         }
     }
 
+    private void addPriorityTask(Task task) {
```
Не совсем так, время начала может и не совпадать.
Но например задача идет с понедельника по среду, а новая задача пришла на вторник, время начала не совпадает, но добавить ее нельзя, ведь вторник занят. 

Спойлер:
```

	private void addPriorityTask(Task task) {
		final LocalDateTime startTime = task.getStartTime();
		final LocalDateTime endTime = task.getEndTime();
		for (Task t : prioritizedTasks) {
			final LocalDateTime existStart = t.getStartTime();
			final LocalDateTime existEnd = t.getEndTime();
			if (!endTime.isAfter(existStart)) {// newTimeEnd <= existTimeStart
				continue;
			}
			if (!existEnd.isAfter(startTime)) {// existTimeEnd <= newTimeStart
				continue;
			}

			throw new TaskValidationException("Задача пересекаются с id=" + t.getId() + " c " + existStart + " по " + existEnd);
		}

		prioritizedTasks.add(task);
	}
```

YazykovMikhail commented on src/ru/yandex/javacource/golotin/schedule/service/InMemoryTaskManager.java:
```diff
@@ -155,13 +160,13 @@ public List<Epic> getEpics() {
+//        }
+        return epic.getSubtaskIds().stream().map(subtasks::get).collect(Collectors.toList());
     }
 
```
Как раз тут должен быть еще метод по подсчету времени Эпика, ведь его время начала/продолжительность/конец зависи от сабтасок. 


СПОЙЛЕР:

```
private void updateEpicDuration(Epic epic) {
		List<Integer> subs = epic.getSubtaskIds();
		if (subs.isEmpty()) {
			epic.setDuration(0L);
			return;
		}
		LocalDateTime start = LocalDateTime.MAX;
		LocalDateTime end = LocalDateTime.MIN;
		long duration = 0L;
		for (int id : subs) {
			final Subtask subtask = subtasks.get(id);
			final LocalDateTime startTime = subtask.getStartTime();
			final LocalDateTime endTime = subtask.getEndTime();
			if (startTime.isBefore(start)) {
				start = startTime;
			}
			if (endTime.isAfter(end)) {
				end = endTime;
			}
			duration += subtask.getDuration();
		}
		epic.setDuration(duration);
		epic.setStartTime(start);
		epic.setEndTime(end);
	}
```

YazykovMikhail replied:
И так как этот метод будет вызываться там же где вызывался updateEpicStatus
то можно сделать еще один метод и обьеденить их в 
```
	protected void updateEpic(int epicId) {
		Epic epic = epics.get(epicId);
		updateEpicStatus(epic);
		updateEpicDuration(epic);
	}
```
И вызывать его, везде, где раньше вызывался updateEpicStatus, ведь если в сабтасках что то меняется, то в и эпике меняется время/статус и нужно их подсчитывать

YazykovMikhail reviewed (changes requested):
Так же есть пункт в задании:
"Покрываем код тестами"
Стоит его выполнить

1EVILGUN1 committed:
Java sprint 8 v1.1

исправлен startTime,duration. Поправлена работа файлового менеджера. Дописал тест к новым методам менеджера. Также добавил класс тест для файлового менеджера.
